### PR TITLE
Re-populating lists with ReadStateAsync

### PR DESCRIPTION
**Pull request by vella-nicholas**
I have encountered an issue when using `ReadStateAsync`. During deserialization, it is appending the list from persistance storage instead of overwriting it. I am including code to replicate it.

Can you confirm this is an issue please?

### CODE REVIEW

File: Orleans.Providers.MongoDB/StorageProviders/Serializers/JsonGrainStateSerializer.cs [hunk 1]
```diff
@@ -23,6 +23,8 @@ protected JsonGrainStateSerializer(JsonSerializer serializer)
             // values that are not equal to the system defaults.
             this.serializer.NullValueHandling = NullValueHandling.Include;
 
+            this.serializer.ObjectCreationHandling = ObjectCreationHandling.Replace;
```
Comment (vella-nicholas) [inline Orleans.Providers.MongoDB/StorageProviders/Serializers/JsonGrainStateSerializer.cs]: This line of code fixes my issue. Try running the code with and without and you will see the difference.
Can you pass a custom `JsonSerializer`? If yes the it makes this change backward compatible (I can clean the test code and merge to master).

File: Test/Grains/EmployeeLeave/VacationLeave.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+namespace Orleans.Providers.MongoDB.Test.Grains
+{
+    public class VacationLeave : IEmployeeLeave
+    {
+        public VacationLeave(int identifier)
+        {
+            Identifier = identifier;
+        }
+
+        public int Identifier { get; set; }
```
Comment (vella-nicholas) [inline Test/Grains/EmployeeLeave/VacationLeave.cs]: This number is only used to identify duplicate records.

File: Orleans.Providers.MongoDB/StorageProviders/Serializers/JsonGrainStateSerializer.cs (hunk 1 above)
Comment (ReubenBond) [reply Orleans.Providers.MongoDB/StorageProviders/Serializers/JsonGrainStateSerializer.cs]: I recommend following the pattern just implemented in the main repo - adding a delegate for configuring the `JsonSerializerSettings`: https://github.com/dotnet/orleans/pull/5662

Comment (vella-nicholas) [reply Orleans.Providers.MongoDB/StorageProviders/Serializers/JsonGrainStateSerializer.cs]: PR has been updated.

**Comment by SebastianStehle** [comment]
I can push these changes to Nuget the next days.

**Comment by vella-nicholas** [comment]
> I can push these changes to Nuget the next days.

I will push a slight cleanup to the test console.

**Pull request merged**